### PR TITLE
PlayerInfo::ActivePlayer() returns the game's PlayerInfo

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Feature:** Allows access to the PlayerInfo via PlayerInfo::ActivePlayer()

## Feature Details
There has long been requests from developers for a way to access the game's PlayerInfo without having to pass it through every function call. The problem with most approaches is forward compatibility. For example, maybe one day we'll want PlayerInfo to display detailed information about a save file, or multiple PlayerInfo objects for remote users, or perhaps just a copy of the current player to manipulate it in complex ways.

No matter what we do to the game design in the future, one thing will always be true: there is one authoritative PlayerInfo object representing the player being played at this moment. Sure, there may be copies, or objects representing remote players, or ones that can be loaded. But only one is actually playing the game.

This PR defines the location where one obtains that PlayerInfo: `PlayerInfo::ActivePlayer()`. The `main()` now gets its PlayerInfo from there.

### Why Not `GameData::ActivePlayer()`?

The design of Endless Sky is for GameData to have nothing specific to one player, but rather generic to all play-throughs of Endless Sky (systems, fleets, global variables). That is one of the objections to past attempts at putting it in GameData.

## Usage Examples
Don't pass PlayerInfo everywhere.

## Testing Done
Made sure the game still ran.

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

File: source/PlayerInfo.h [hunk 1]
```diff
@@ -64,6 +64,10 @@ class PlayerInfo {
 
 
 public:
+	// Get the PlayerInfo object that represents the current player playing the game.
+	// If no player has been selected yet, this will return an empty PlayerInfo()
```
Comment (Koranir) [inline source/PlayerInfo.h:68]:
```suggestion
	// If no player has been selected yet, this will return an empty PlayerInfo().
```

**Review by Koranir — APPROVED** [review]
I like it!

**Review by quyykk — COMMENTED** [review]
I would rather have this in GameData with the rest. :smile:

**Comment by UnorderedSigh** [comment]
> I would rather have this in GameData with the rest

The design of Endless Sky is for GameData to have anything not specific to the player, but rather generic to all play-throughs of Endless Sky (systems, fleets, global variables). That is one of the objections to past attempts at putting it in GameData.

**Comment by UnorderedSigh** [comment]
> I would rather have this in GameData with the rest

I've added a note about that in the PR description

File: source/PlayerInfo.h [hunk 2]
```diff
@@ -64,6 +64,10 @@ class PlayerInfo {
 
 
 public:
+	// Get the PlayerInfo object that represents the current player playing the game.
+	// If no player has been selected yet, this will return an empty PlayerInfo().
```
Comment (UnorderedSigh) [inline source/PlayerInfo.h:68]:
```suggestion
	// If no player has been selected yet, this will return an empty PlayerInfo.
```

**Comment by UnorderedSigh** [comment]
> I would rather have this in GameData with the rest.

If moving this to GameData will get the PR merged, I'm happy to do that. Either location seems okay to me.

**Comment by UnorderedSigh** [comment]
Last tested and known to work Sun, 16 Apr 2023 23:29:27 UTC

**Comment by UnorderedSigh** [comment]
The Ubuntu 20 test failed due to a sound card issue. My changes have nothing to do with that.

**Comment by UnorderedSigh** [comment]
Last tested and known to work Fri, 12 May 2023 04:53:20 UTC

**Comment by UnorderedSigh** [comment]
Last tested and known to work Mon, 29 May 2023 16:50:43 UTC

File: source/PlayerInfo.h [hunk 3]
```diff
@@ -355,6 +359,8 @@ class PlayerInfo {
 
 
 private:
+	static PlayerInfo activePlayer;
```
Comment (RisingLeaf) [inline source/PlayerInfo.h:362]: Do we normally use static variables in classes? I think usually something like this goes into the unnamed namespace in the cpp.

File: source/PlayerInfo.h [hunk 4]
```diff
@@ -64,6 +64,10 @@ class PlayerInfo {
 
 
 public:
+	// Get the PlayerInfo object for the player that is presently playing the game.
+	// If no player has been selected yet, this will return an empty PlayerInfo.
+	static PlayerInfo &ActivePlayer();
```
Comment (RisingLeaf) [inline source/PlayerInfo.h:69]: This feels like an odd placement in the class, not sure how to improve though.

**Review by RisingLeaf — COMMENTED** [review]
The general idea definitely has my approval, passing the player around is really annoying.

I don't really see it inside of the PlayerInfo class though.
I do agree that it doesn't fit GameData, maybe it needs something new? That might be a little too much on the other hand.

File: source/PlayerInfo.h (hunk 3 above)
Comment (UnorderedSigh) [reply source/PlayerInfo.h:362]: The source is inconsistent on static vs. local. In this case, static makes more sense. It provides a promise that the `Player &` from ActivePlayer() will return something that'll never be deallocated.

**Comment by UnorderedSigh** [comment]
Last tested and known to work Sun, 27 Aug 2023 14:49:22

**Comment by UnorderedSigh** [comment]
Last tested and known to work Wed, 22 Nov 2023 23:55:16 UTC

**Comment by petervdmeer** [comment]
I understand the convenience, but I personally prefer the current "pass the objects as parameters approach", and I feel that we should focus more on decoupling (only providing classes the data that they really need).

An example is the Ship class; CanSendHail() and Scan() are the functions that drag a dependency on PlayerInfo into this class. CanSendHail() wants to know the player-system and the players languages. Scan() just wants to know the players flagship. Losing the dependency on PlayerInfo from Ship would make refactoring of this class a bit easier.

Another example the Planet class; PlayerInfo only gets used for DemandTribute(), and it only needs to have information about current tribute and about setting tribute. Having a dependency on PlayerInfo also seems undesired here. (We might want to check the tribute statusses on the caller, not here?)

ShipInfoDisplay is the 3d class I checked; it wants to know the licenses the player has, deprecation (financial) data, and the current date. PlayerInfo contains a lot more data, but other information from PlayerInfo is not relevant for this ShipInfoDisplay.

I'm convinced that if we focus more on decoupling, then the code becomes easier to test (less dependencies, less work when mocking objects), faster to compile and easier to refactor.

So yes; passing `PlayerInfo player` around at many locations is a bit of a problem, but I'm affraid that making it easier to have `PlayerInfo player` available at even more locations will make some structural changes more difficult in the future.

On a related note: Having multiple PlayerInfo objects seems unlikely, although we already have had a [PR where we have 2 active players on a single computer (side-by-side play)](https://github.com/endless-sky/endless-sky/pull/5578).

**Comment by petervdmeer** [comment]
Adding the `unlikely` label based on my previous comment; the problem is very valid, but this might not be the solution.

**Pull request closed without merging**